### PR TITLE
[FSTORE-980] Helper, primary key and event time columns with feature view

### PR DESCRIPTION
**Pull request by davitbzh**
This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/feature_view.py [hunk 1]
```diff
@@ -2097,6 +2344,9 @@ def get_train_test_split(
         self,
         training_dataset_version,
         read_options: Optional[Dict[Any, Any]] = None,
+        primary_keys=False,
+        event_time=False,
```
Comment (kennethmhc) [inline python/hsfs/feature_view.py:2348]: why they are needed when there is already `training_helper_columns`?
shouldn't users add primary key and event time to the helper columns?

File: python/hsfs/core/vector_server.py [hunk 2]
```diff
@@ -232,45 +394,24 @@ def get_feature_vector(
...
-        else:
-            raise Exception(
-                "Unknown return type. Supported return types are 'list', 'pandas' and 'numpy'"
-            )
-
-    def get_feature_vectors(
-        self, entries, return_type=None, passed_features=[], allow_missing=False
-    ):
-        """Assembles serving vector from online feature store."""
+        return result_dict, serving_vector
```
Comment (kennethmhc) [inline python/hsfs/core/vector_server.py:397]: `result_dict` only stores the result of the last prepared statement. same for the batch case.

File: python/hsfs/core/vector_server.py [hunk 3]
```diff
@@ -191,10 +236,125 @@ def _validate_serving_key(self, entry):
...
+        self,
+        entries,
+        return_type,
+    ):
+        """Assembles serving vector from online feature store."""
+        batch_results, serving_keys = self._batch_vector_results(
+            entries, self._helper_column_prepared_statements
+        )
+
+        # drop serving key names
```
Comment (kennethmhc) [inline python/hsfs/core/vector_server.py:340]: why drop serving key here even though they are requested by users?

File: python/hsfs/core/vector_server.py (hunk 2 above)
Comment (davitbzh) [reply python/hsfs/core/vector_server.py:397]: I am not sure I understood. I just put this in a function to not duplicate the code. It must do the same as here https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/core/vector_server.py#L248 

Am I missing something?

File: python/hsfs/feature_view.py [hunk 4]
```diff
@@ -514,6 +523,112 @@ def get_feature_vectors(
...
+            Returned `pd.DataFrame` or `List[dict]`  contains feature values related to provided primary
+            keys, ordered according to positions of this features in the feature view query.
+
+        # Raises
+            `Exception`. When primary key entry cannot be found in one or more of the feature groups used by this
+                feature view.
+        """
+        if (
+            self._batch_vectors_server is None
+            or self._single_vector_server._helper_column_prepared_statements is None
```
[the remark this answers is not in the document]
Comment (davitbzh) [reply python/hsfs/feature_view.py]: done

File: python/hsfs/core/vector_server.py (hunk 2 above)
Comment (kennethmhc) [reply python/hsfs/core/vector_server.py:410]: [here](https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/core/vector_server.py#L248)
`result_dict` store the value of transformed `serving_vector`
`result_dict = self._apply_transformation(serving_vector)`

but in the code above 
```
for row in result_proxy:
                    result_dict = self.deserialize_complex_features(
                        self._complex_features, row._asdict()
                    )
                    serving_vector.update(result_dict)
```
the last assignment to result_dict is inside a for loop.

Comment (davitbzh) [reply python/hsfs/core/vector_server.py:410]: yes but `_vector_result` is just copy of from [here](https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/core/vector_server.py#L214) to [here](https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/core/vector_server.py#L243). That was the whole idea. I am still unable to spot difference

File: python/hsfs/core/vector_server.py [hunk 5]
```diff
@@ -315,14 +319,14 @@ def get_feature_vectors(
     def get_inference_helper(self, entry, return_type):
         """Assembles serving vector from online feature store."""
 
-        result_dict, _ = self._vector_result(
+        serving_vector = self._vector_result(
```
Comment (kennethmhc) [inline python/hsfs/core/vector_server.py:322]: need to filter out features which are not helper columns?

**Review by kennethmhc — APPROVED** [review]
LGTM!

**Comment by davitbzh** [comment]
Trigger

**Comment by davitbzh** [comment]
Trigger

**Pull request merged**